source/optional/match.d

```diff
@@ -15,38 +15,71 @@ import optional.optional;
+	        } else static if (is(typeof(handlers[0]()))) {
+	            alias someHandler = handlers[1];
+	            alias noHandler = handlers[0];
+				return doMatch!(someHandler, noHandler, O, T)(opt);
+	        } else {
+				// One of these two is causing a compile error.
+				// Let's call them so the compiler can show a proper error warning.
+				failOnCompileError!(handlers[0], T);
+				failOnCompileError!(handlers[1], T);
+			}
+		} else static if (is(typeof(opt.value) == Optional!T, T)) {
+			return opt.valueMatch!handlers;
+		} else {
+			pragma(msg, "Type of: " ~ typeof(opt.value).stringof);
+			static assert(0, "Cannot match!() on a " ~ O.stringof);
+		}
+	}
+}
 
-        static if (is(typeof(handlers[0](opt.front)))) {
-            alias someHandler = handlers[0];
-            alias noHandler = handlers[1];
-        } else {
-            alias someHandler = handlers[1];
-            alias noHandler = handlers[0];
-        }
+private auto doMatch(alias someHandler, alias noHandler, O, T)(ref auto O opt) {
+	alias SomeHandlerReturn = typeof(someHandler(T.init));
+	alias NoHandlerReturn = typeof(noHandler());
+	enum isVoidReturn = is(typeof(someHandler(T.init)) == void) || is(typeof(noHandler()) == void);
 
-        alias SomeHandlerReturn = typeof(someHandler(opt.front));
-        alias NoHandlerReturn = typeof(noHandler());
-        enum isVoidReturn = is(SomeHandlerReturn == void) || is(NoHandlerReturn == void);
+	static assert(
+		is(SomeHandlerReturn == NoHandlerReturn) || isVoidReturn,
+		"Expected two handlers to return same type, found type '" ~ SomeHandlerReturn.stringof ~ "' and type '" ~ NoHandlerReturn.stringof ~ "'",
+	);
 
-        static assert(
-            is(SomeHandlerReturn == NoHandlerReturn) || isVoidReturn,
-            "Expected two handlers to return same type, found type '" ~ SomeHandlerReturn.stringof ~ "' and type '" ~ NoHandlerReturn.stringof ~ "'",
-        );
+	if (opt.empty) {
+		static if (isVoidReturn) {
+			noHandler();
+		} else {
+			return noHandler();
+		}
+	} else {
+		static if (isVoidReturn) {
+			someHandler(opt.front);
+		} else {
+			return someHandler(opt.front);
+		}
+	}
+}
 
-        if (opt.empty) {
-            static if (isVoidReturn) {
-                noHandler();
-            } else {
-                return noHandler();
-            }
-        } else {
-            static if (isVoidReturn) {
-                someHandler(opt.front);
-            } else {
-                return someHandler(opt.front);
-            }
-        }
+/**
+Performs a match on the `value` property of a variable.
+*/
+private template valueMatch(handlers...) if (handlers.length == 2) {
+	auto valueMatch(O)(auto ref O opt) {
+		return opt.value.match!handlers;
+	}
+}
+
+/**
+Prints out the correct compile error if the handler cannot be compiled.
+*/
+private void failOnCompileError(alias handler, T)() {
+	static if (!is(typeof(handler(T.init))) && !is(typeof(handler()))) {
+		cast(void) handler(T.init);
```

Comment:
Neat trick!

Reply:
Thanks! Had to try out some stuff to get those errors out properly.